src/main/java/pl/sda/transporeon/currencyexchange/service/ExchangeRateService.java

```diff
@@ -2,46 +2,78 @@
 import org.springframework.stereotype.Service;
-import pl.sda.transporeon.currencyexchange.model.ExchangeRate;
-import pl.sda.transporeon.currencyexchange.model.ExchangeRateDTO;
+import org.springframework.web.client.HttpClientErrorException;
+import pl.sda.transporeon.currencyexchange.controller.exception.RateProcessingException;
+import pl.sda.transporeon.currencyexchange.model.*;
 import pl.sda.transporeon.currencyexchange.repository.ExchangeRateRepository;
 
+import java.time.LocalDate;
+import java.util.List;
+
+import static pl.sda.transporeon.currencyexchange.controller.ExchangeRateController.gold;
+
 @Service
 public class ExchangeRateService {
 
     private final ExchangeRateRepository exchangeRateRepository;
-    private final ApiServiceNonPln apiServiceNonPln;
-    private final ApiServiceGold apiServicePln;
-    private  final MapToDTO mapToDTO;
+    private final RestTemplateConfig restTemplate;
+    private final ExchangeRateMapper mapper;
 
     @Autowired
-    public ExchangeRateService(ExchangeRateRepository exchangeRateRepository, ApiServiceNonPln apiServiceNonPln, ApiServiceGold apiServicePln, MapToDTO mapToDTO) {
+    public ExchangeRateService(ExchangeRateRepository exchangeRateRepository,  RestTemplateConfig restTemplate, ExchangeRateMapper mapper) {
         this.exchangeRateRepository = exchangeRateRepository;
-        this.apiServiceNonPln = apiServiceNonPln;
-        this.apiServicePln = apiServicePln;
-        this.mapToDTO = mapToDTO;
+        this.restTemplate = restTemplate;
+        this.mapper = mapper;
     }
 
 
-    public ExchangeRateDTO getExchangeDataToView(String base,
-                                                 String target,
-                                                 String date){
+    public ExchangeRateDTO getExchangeDataToView(String base, String target, String date) {
         ExchangeRate rate;
-        if(base.equals("gold")){
-             rate = apiServicePln.getRate(base, target, date);
-        }else {
-             rate = apiServiceNonPln.getRate(base, target, date);
+        String request;
+        boolean recordExists = doesRecordExists(base, target, mapper.stringDateToLocalDate(date));
+
+        try {
+            if (!recordExists) {
+                if(base.equals(gold)){
+                    request = "http://api.nbp.pl/api/cenyzlota/" + date + "/";
+                    rate = mapper.mapGold(getGoldRate(request));
+                } else {
+                    request = "https://api.exchangerate.host/" + date + "?base=" + base + "&symbols=" + target;
+                    rate = mapper.mapCurrency(getCurrencyRate(request), target);
+                }
+                exchangeRateRepository.save(rate);
+            } else {
+                List<ExchangeRate> records = findRecord(base, target, mapper.stringDateToLocalDate(date));
+                rate = records.get(0);
+            }
+        } catch (HttpClientErrorException | NullPointerException e) {
```

Comment:
Dobra praktyka mówi - nie łapiemy NPE (NullPointerException), tylko je minimalizujemy, dostosowujemy kod, żeby nie leciały (obsługujemy to kodem)